File: leetcode/leetcode_cpp/design-search-autocomplete-system.cpp
#include <iostream>
#include <cassert>
#include <vector>
#include <queue>
#include <unordered_map>

using namespace std;

/**
https://leetcode.com/problems/design-search-autocomplete-system/

//Solution1. Build Trie. And BSF to save autocomplete list in cache
//time for build trie: o(N). build trie(n*L)L:length of sentence
//time for autocomplete: o(N+M). N:depth to cur node. M: all decendent children
//space: o(N+M). N: trie, M: recursion

*/

class Trie {
public:
    Trie() {
        needToUpdate = true;
    }
    
    ~Trie() {
        for (auto &n:children) delete n.second;
    }
    
    void addWord(string &s, int time) {
        auto trie = this;
        for (auto &c:s) {
            if (trie->children[c] == nullptr) trie->children[c] = new Trie();
            trie = trie->children[c];
            //trie->counter[s] += time;
            trie->needToUpdate = true;
        }
        trie->s = s;
        trie->count += time;
    }
    
    struct maxHeap_comp{
        bool operator() (const pair<string,int> &a, const pair<string,int> &b) {
            return (a.second == b.second) ? a.first > b.first : a.second < b.second;
        }
    };
    priority_queue<pair<string, int>, vector<pair<string, int>>, maxHeap_comp> q;
    
    void buildHeap(Trie* node) { // BFS
        if (node == this) q = priority_queue<pair<string, int>, vector<pair<string, int>>, maxHeap_comp>();
        if (node && node->s != "") {
            q.push({node->s, node->count});
        }
        for (auto &child:node->children) {
            buildHeap(child.second);
        }
    }
    
    vector<string> autoCompletation() {
        if (needToUpdate) {
            buildHeap(this);

            cache.clear();
            while (!q.empty() && cache.size() < 3) {
                cache.push_back(q.top().first);
                q.pop();
            }
            needToUpdate = false;
        }
        return cache;
    }
    
    unordered_map<char, Trie*> children;
    string s;
    int count;
    bool needToUpdate;
    
    //unordered_map<string,int> counter;
    
    vector<string> cache;
};

class AutocompleteSystem {    
public:
    AutocompleteSystem() {
        root = nullptr;
        cur = nullptr;
    }
    
    ~AutocompleteSystem() {
        if (root) delete root;
    }
    
    AutocompleteSystem(vector<string>& sentences, vector<int>& times) {
        root = new Trie();
        for (int i=0; i<sentences.size(); ++i) {
            root->addWord(sentences[i], times[i]);
        }
        cur = root;
    }
    
    vector<string> input(char c) {
        if (c == '#') {
            cur = root;
            root->addWord(keyword, 1);
            keyword = "";
            return {};
        }
        keyword += c;
        if (cur != NULL) cur = cur->children[c];
        return (cur != NULL) ? cur->autoCompletation() : vector<string>();
    }
    Trie *root;
    Trie *cur;
    string keyword;
};

/**
 * Your AutocompleteSystem object will be instantiated and called as such:
 * AutocompleteSystem* obj = new AutocompleteSystem(sentences, times);
 * vector<string> param_1 = obj->input(c);
 */

int main()
{
    auto input1_1 = vector<string> {"i love you","island","iroman","i love leetcode"};
    auto input1_2 = vector<int> {5,3,2,2};
    auto autoCompleteSystem = AutocompleteSystem(input1_1, input1_2);
    assert(autoCompleteSystem.input('i') == (vector<string>{"i love you","island","i love leetcode"}));
    assert(autoCompleteSystem.input(' ') == (vector<string>{"i love you","i love leetcode"}));
    assert(autoCompleteSystem.input('a') == (vector<string>{}));
    assert(autoCompleteSystem.input('#') == (vector<string>{}));
    assert(autoCompleteSystem.input('i') == (vector<string>{"i love you","island","i love leetcode"}));
    assert(autoCompleteSystem.input(' ') == (vector<string>{"i love you","i love leetcode","i a"}));
    assert(autoCompleteSystem.input('a') == (vector<string>{"i a"}));
    assert(autoCompleteSystem.input('#') == (vector<string>{}));
    assert(autoCompleteSystem.input('i') == (vector<string>{"i love you","island","i a"}));
    assert(autoCompleteSystem.input(' ') == (vector<string>{"i love you","i a","i love leetcode"}));
    assert(autoCompleteSystem.input('a') == (vector<string>{"i a"}));
    assert(autoCompleteSystem.input('#') == (vector<string>{}));

    return 0;
}

File: leetcode/leetcode_python/design-search-autocomplete-system.py
from typing import List
import collections
import heapq

"""
https://leetcode.com/problems/design-search-autocomplete-system/submissions/
"""

class Trie:
    def __init__(self):
        self.children = {}
        self.sentence = ''
        self.count = 0
        #self.needToUpdate = True
    
    def addWord(self, s: str, count: int):
        trie = self
        for c in s:
            if c not in trie.children:
                trie.children[c] = Trie()
            trie = trie.children[c]
            #trie.needToUpdate = True
            
        trie.sentence = s
        trie.count += count
        
    def buildAutoCompletation(self, q, node):
        if node and node.count > 0:
            q.append((node.count, node.sentence))
            
        for k,v in node.children.items():
            self.buildAutoCompletation(q, v)
        
    def autoCompletation(self):
        #if self.needToUpdate:
        q = []
        self.buildAutoCompletation(q, self)
        result=heapq.nsmallest(3,q,key= lambda x:(-x[0],x[1])) # max heap to get top 3 
        return [word[1] for word in result]
        #return self.cache
    

class AutocompleteSystem:

    def __init__(self, sentences: List[str], times: List[int]):
        self.root = Trie()
        self.keyWord = ''
        for i in range(0, len(sentences)):
            self.root.addWord(sentences[i], times[i])
        self.cur = self.root

    def input(self, c: str) -> List[str]:
        if c == '#':
            self.root.addWord(self.keyWord,1)
            self.cur = self.root
            self.keyWord = ''
            return []
        
        self.keyWord += c
        if self.cur:
            self.cur = self.cur.children[c] if c in self.cur.children else None
        
        if self.cur:
            return self.cur.autoCompletation()
        else:
            return []


# Your AutocompleteSystem object will be instantiated and called as such:
# obj = AutocompleteSystem(sentences, times)
# param_1 = obj.input(c)